docs/v3/core/v3.0.rst

```diff
@@ -1089,6 +1089,36 @@ each of these two representations are defined to be either:
 - a multi-dimensional array of some shape and data type, or
 - a byte string.
 
+Based on the input and output representations for the encode transform,
+codecs can be classified as one of three kinds:
+
+- ``array -> array``
+- ``array -> bytes``
+- ``bytes -> bytes``
+
+.. note::
+
+   ``bytes -> array`` codecs, where after encoding an array as a byte
+   string, it is subsequently transformed back into an array, to then later
+   be transformed back into a byte string, are not currently allowed, due to
+   the lack of a clear use case.
+
+If multiple codecs are specified for an array, each codec is applied
+sequentially; when encoding, the encoded output of codec ``i`` serves as the
+decoded input of codec ``i+1``, and similarly when decoding, the decoded output
+of codec ``i+1`` serves as the encoded input to codec ``i``.  Since ``bytes ->
+array`` codecs are not supported, it follows that the list of codecs must be of
+the following form:
+
+- zero or more ``array -> array`` codecs; followed by
+- at most one ``array -> bytes`` codec; followed by
+- zero or more ``bytes -> bytes`` codecs.
+
+If no ``array -> bytes`` codec is specified, then the default byte
+representation for the data type of the array is used.  For all data types
+currently defined by the core spec, that is equivalent to the ``endian`` codec
+with an endianness of ``little``.
```

Comment:
Should we also mention the `transpose` codec with order `C`?

Reply:
The endian codec with endianness of little may be a noop in terms of the actual in-memory representation of the data, but it does have an effect in terms of the logical data model, in that it transforms a multidimensional array into a (one-dimensional) byte sequence.  In contrast, the transpose codec with order [0, 1, ..., rank-1], i.e. "C", is logically an identity transform.  It is the endian codec that uses C order, and if you want a different order you have to transpose it first.